lib/sync.js

```diff
@@ -481,17 +481,31 @@ SyncApi.prototype._processRoomEvents = function(room, stateEventList,
 };
 
 function retryTimeMsForAttempt(attempt) {
-    // 2,4,8,16,32,64,128,128,128,... seconds
-    // max 2^7 secs = 2.1 mins
-    return Math.pow(2, Math.min(attempt, 7)) * 1000;
+    // 2,4,8,16,32,32,32,32,... seconds
+    // max 2^5 secs = 32 secs
+    return Math.pow(2, Math.min(attempt, 5)) * 1000;
 }
 
 function startSyncingRetryTimer(client, attempt, fn) {
     client._syncingRetry = {};
     client._syncingRetry.fn = fn;
+    var newAttempt = attempt;
+    var timeBeforeWaitingMs = Date.now();
+    var timeToWaitMs = retryTimeMsForAttempt(attempt);
     client._syncingRetry.timeoutId = setTimeout(function() {
-        fn();
-    }, retryTimeMsForAttempt(attempt));
+        var timeAfterWaitingMs = Date.now();
+        var timeDeltaMs = timeAfterWaitingMs - timeBeforeWaitingMs;
+        if (timeDeltaMs > (2 * timeToWaitMs)) {
+            // we've waited more than twice what we were supposed to. Reset the
```

Comment:
Should say why this could ever happen - which is basically that the comp went to sleep, or the clock was stepped.

The 2x factor feels a bit weird.  I'd have thought +/- 10% difference in what we expected should be reason to be suspicious and reset the retry schedule.